services/121-service/src/notifications/message-incoming/message-incoming.service.ts

```diff
@@ -372,31 +372,41 @@ export class MessageIncomingService {
     });
 
     const filteredRegistrations: RegistrationEntity[] = [];
-    for (const r of registrationWithVouchers) {
+    for (const registration of registrationWithVouchers) {
       // Don't send more then 3 vouchers, so no vouchers of more than 2 payments ago
-      const lastPayment = await this.transactionRepository
-        .createQueryBuilder('transaction')
-        .select('MAX(transaction."paymentId")', 'max')
-        .leftJoin('transaction.payment', 'p')
-        .where('p.programId = :programId', {
-          programId: r.programId,
-        })
-        .getRawOne();
-      const minimumPayment = lastPayment ? lastPayment.max - 2 : 0;
-
-      r.images = r.images.filter(
+      const lastThreePaymentIds =
+        await this.getLastThreePaymentIdsForRegistration(registration.id);
+
+      registration.images = registration.images.filter(
         (image) =>
           !image.voucher.send &&
           image.voucher.paymentId &&
-          image.voucher.paymentId >= minimumPayment,
+          lastThreePaymentIds.includes(image.voucher.paymentId),
```

Comment:
This filter could have been done in the query. I know this was always already the case that we filtered here before your refactor,  but it's probably not great for performance to load a lot of images in memory. 

But we of course never had issues with this part before so up to you if you want to change it.

Reply:
opted not to do this. This code is also about just one registration, so performance will not quickly be a problem I think.